### ISSUE TITLE
Walrus operator + type inference + generic class + some other conditions = mypy failure

### ISSUE THREAD

**Issue by dnaaun** [python/mypy#9743]
**Bug Report**
```python
from typing import (
    Generic,
    TypeVar,
)
_MyType = TypeVar("_MyType", str, float) 

class MyGeneric(Generic[_MyType]):
    def __init__(self, attr: int) -> None:
        ls = [0]
        self.attr = attr          # mypy says: error: Need type annotation for 'attr'
        if (a := sum(ls)): # mypy says: error: Cannot determine type of 'ls'
            pass
        reveal_type(ls)        # mypy says: note: Revealed type is 'builtins.list[builtins.int*]'
```

The title of this bug report was a bit vague because I didn't know how to concisely describe the necessary set of conditions for this bug. Here are those conditions:

1.  `ls` is not manually annotated.
2. The walrus operator is used.
4. The block is inside a generic class.
5. `_MyType` is a TypeVar that explicitly enumerates all the possible types (as opposed to being `bound=...` or just a free TypeVar)

**Environment**
mypy 0.800+dev.40fd841a005936d95e7c351435f1b7b7b4bdd43d
No mypy.ini config, no cmdline flags.
Python 3.8.3

**Comment by GBeauregard** [comment]
I believe I've ran into this same bug when converting one of my classes to a generic class as part of a rewrite. You can find information on gitter here: https://gitter.im/python/typing?at=6158b2c638377967f42dba74

Here is the failing example:
https://gist.github.com/GBeauregard/9edc42fb7e855d65ea89c756ccd82665

I ran into this when converting the class from using `str` to being a generic class of `AnyStr`. I've left it as `str` for now before I decide what to do when rewriting the class.

**Comment by GBeauregard** [comment]
I can confirm this still affects the latest git tag `mypy 0.920+dev.0a830481980bfc554ded61a3eaaaecde384a21e4`.

**Comment by erictraut** [comment]
This bug appears to have been fixed. I cannot repro it with the latest version of mypy.

**Comment by hauntsaninja** [comment]
Fixed in #14135
